vir/src/macros.rs

```diff
@@ -304,215 +312,233 @@ macro_rules! vir_predicate {
-        $crate::expr!(@forall_qvars($output, $qvars); $($tokens)*)
-    } };
-    (@forall_qvars($output:ident, $qvars:ident); $qvar:ident : $qtype:tt $($tokens:tt)* ) => { {
-        let local = vcx!().mk_local(stringify!($qvar), $crate::expr!(@typ; $qtype));
-        $qvars.push(vcx!().mk_local_decl_local(local));
-        let $qvar: $crate::Expr = vcx!().mk_local_ex_local(local);
-        $crate::expr!(@forall_qvars($output, $qvars); $($tokens)*)
+    ($vcx:tt; $($tokens:tt)*) => { {
+        use $crate::macros::{ExprApply, ExprQuote};
+        let vcx = $vcx; #[allow(unused)] macro_rules! vcx { () => { vcx }; }
+        $crate::expr_inner!(@expr_one; $($tokens)*)
     } };
-    (@forall_qvars($output:ident, $qvars:ident); , $($tokens:tt)* ) => { // TODO: this accepts too many commas
-        $crate::expr!(@forall_qvars($output, $qvars); $($tokens)*)
-    };
-    (@forall_qvars($output:ident, $qvars:ident); ) => { compile_error!("malformed forall") };
-
-    (@expr_done($output:ident); , $($tokens:tt)+) => { $crate::expr!(@expr($output); $($tokens)*) };
-    (@expr_done($output:ident); $($tokens:tt)+) => { compile_error!("unexpected VIR expression (missing comma?)") };
-    (@expr_done($output:ident);) => {};
+    ($($tokens:tt)+) => { vir::with_vcx(|vcx| $crate::expr!(vcx; $($tokens)*)) };
+    () => { compile_error!(concat!("VIR malformed empty")) };
+}
 
-    (@expr($output:ident); unfolding_wildcard ( [ $outer:expr ]( $($args:tt)* ) ) in ( $($rhs:tt)+ ) ) => { { $output.push(vcx!().mk_unfolding_expr(
-        $outer.apply(vcx!(),
-            $crate::expr!(@expr_list; $($args)*).as_slice(),
-            Some(vcx!().mk_wildcard()),
-        ),
-        $crate::expr!(@expr_one; $($rhs)*),
-    )); } };
-    (@expr($output:ident); acc( [ $outer:expr ]( $($args:tt)* ) ) ) => { { $output.push(vcx!().mk_predicate_app_expr(
-        $outer.apply(vcx!(),
-            $crate::expr!(@expr_list; $($args)*).as_slice(),
-            None,
+#[macro_export]
+macro_rules! expr_inner {
+    (@expr_one; ($($e:tt)+) as $ty:ident) => {
+        $crate::CastType::inner_cast_ty::<$crate::$ty>($crate::expr_inner!(@expr_one; $($e)*))
+    };
+    (@expr_one; unfolding ( [ $outer:expr ] ( $($args:tt)* ) ) in ( $($rhs:tt)+ ) ) => { vcx!().mk_unfolding_expr(
+        $outer.call($crate::expr_inner!(@expr_args; $($args)*))(None),
+        $crate::expr_inner!(@expr_one; $($rhs)*),
+    ) };
+    (@expr_one; acc( [ $outer:expr ] ( $($args:tt)* ) ) ) => { {
+        let pred: $crate::PredicateIdn<_> = $outer;
+        pred.expr_apply(
+            vcx!(),
+            $crate::expr_inner!(@expr_args; $($args)*),
         )
-    )); } };
-    (@expr($output:ident); acc_field( [ $outer:expr ]( $($args:tt)* ) ) ) => { { $output.push(vcx!().mk_acc_field_expr(
-        $crate::expr!(@expr_one; $($args)*),
+    } };
+    (@expr_one; acc_field( ( $($args:tt)+ ).[ $outer:expr ] ) ) => { vcx!().mk_acc_field_expr(
```

Comment:
With this change, could it also be `acc` instead of `acc_field` without Rust complaining about an ambiguity?